Normalize the constellation points so that the mean square magnitude is 1 (instead of the non-squared magnitude).

eduardosm opened the pull request:
This allows to easily create signals with normalized power and makes it consistent with the 16-QAM constellation (which is also normalized for mean square magnitude = 1).

marcusmueller commented:
Hi! Thanks for contributing this PR (I like it, especially since you've even thought of modifying the documenation).

This of course breaks the unit tests, so unless you also adapt these, we can't merge this. Also, it changes the way an existing block that people rely on working the way it does – that's a no-go, I'm afraid.

I can see your use case, though! Power normalization actually *is* the more common way of normalization than amplitude normalization. So, what about the following:

1. change the `normalize_points` parameter to not be a `bool`, but an `enum` that allows choosing between `NO_NORMALIZATION`, `POWER_NORMALIZATION`, `AMPLITUDE_NORMALIZATION`
2. use that parameter to pick the appropriate normalization method
3. (important) add a unit test that tests all three modes, including the new power normalization mode.

marcusmueller commented:
Oh, and while you can possibly fly under the radar with this "small change" (it could boil down to < 10 LOC) in terms of copyright: I think it would be wisest if you contacted bhilburn AT gnuradio PERIOD org (@bhilburn) and got a Contributor license agreement, which allows the FSF to defend the GPL code, and allows you to do whatever you want with your own code.

eduardosm commented:
>     1. change the `normalize_points` parameter to not be a `bool`, but an `enum` that allows choosing between `NO_NORMALIZATION`, `POWER_NORMALIZATION`, `AMPLITUDE_NORMALIZATION`
> 
>     2. use that parameter to pick the appropriate normalization method

I have pushed a new commit implementing it as suggested.

>     3. (important) add a unit test that tests all three modes, including the new power normalization mode.

Tests are now passing, because I set the default to `AMPLITUDE_NORMALIZATION`, but I am not sure how to extend the current tests to cover all the cases.

I have also modified the GRC block definition (`digital_constellation.block.yml`) to allow choosing the normalization type for 'Variable Constellation', but I haven't had a chance to test it yet.



> Oh, and while you can possibly fly under the radar with this "small change" (it could boil down to < 10 LOC) in terms of copyright: I think it would be wisest if you contacted bhilburn AT gnuradio PERIOD org (@bhilburn) and got a Contributor license agreement, which allows the FSF to defend the GPL code, and allows you to do whatever you want with your own code.

Thanks for the tip, I have sent an email to get more information.

jdemel commented:
This should fix #2764 . I have a QA test in `gr-gfdm` that relies on normalized amplitudes in the constellation object. It did work with 3.7 and broke in 3.8. I'd argue that 3.8 actually introduced a regression here. Thus, making normalized amplitudes the default would actually NOT break other things.

michaelld commented:
@eduardosm @bhilburn did we get your CLA finalized?

mormj commented:
@eduardosm  - we'd love to be able to include this PR - any update on the CLA process?

eduardosm commented:
I have been asked to get a disclaimer for my university. Due to the current situation here in Spain, I don't know when that will be possible.

Anyways, is the CLA that important for this PR? It is not like I have invented the wheel here...

michaelld commented:
> Anyways, is the CLA that important for this PR? It is not like I have invented the wheel here...

Yes the CLA is required, since the changes are more than ~10 lines of code. Even much more simple changes can require a CLA if there are enough changes ...

eduardosm commented:
> Yes the CLA is required, since the changes are more than ~10 lines of code. Even much more simple changes can require a CLA if there are enough changes ...

This will have to wait then. This CLA thing feels to me like a barrier to entry for new contributors...

mohmmadd commented:
> > Yes the CLA is required, since the changes are more than ~10 lines of code. Even much more simple changes can require a CLA if there are enough changes ...
> 
> This will have to wait then. This CLA thing feels to me like a barrier to entry for new contributors...

@eduardosm Are you have any good news?

mormj commented:
@eduardosm - We have removed some of the barrier for new contributors by using the DCO instead of CLA (https://github.com/gnuradio/gnuradio/blob/master/CONTRIBUTING.md)

All this means, is that we can merge your changes if you signoff your commit git commit --amend -s and force push

(looks like it needs a rebase as well)

eduardosm commented:
> @eduardosm - We have removed some of the barrier for new contributors by using the DCO instead of CLA (https://github.com/gnuradio/gnuradio/blob/master/CONTRIBUTING.md)
> 
> All this means, is that we can merge your changes if you signoff your commit git commit --amend -s and force push
> 
> (looks like it needs a rebase as well)

Thanks for the notification. I have rebased and added the signoff as suggested.


Build is failing with
```
CMake Error at cmake/Modules/GrPybind.cmake:86 (message):
  Python bindings for constellation.h are out of sync
Call Stack (most recent call first):
  gr-digital/python/digital/bindings/CMakeLists.txt:97 (GR_PYBIND_MAKE_CHECK_HASH)
```

But I am not sure where those bindings are located. Aren't those generated automatically?

mormj commented:
> But I am not sure where those bindings are located. Aren't those generated automatically?

Unfortunately no - I can help with this if you would like because it would be nice to have this feature ready for the 3.9 release cutoff.  The python bindings are located in the python/bindings/*_python.cc files and follow standard pybind11 syntax.  There is a helper script bind_intree_file.py that might be able to help.

eduardosm commented:
I think I've got the bindings

marcusmueller commented:
@eduardosm Awesome!
